### PR TITLE
Fix temporary pushing with CSE

### LINKED ISSUE #74: Bad code motion with separately specified quad degrees
Consider:

```python
from firedrake import *
# parameters["coffee"] = {}

m = UnitSquareMesh(1, 1)

V = VectorFunctionSpace(m, 'DG', 1)
u = Function(V)
v = TestFunction(V)

n = FacetNormal(m)

F = (
    + inner(outer(v, n), (outer(u, n)))*ds(degree=1) \
    - inner(grad(v), (outer(u, n)))*ds(degree=3)
)

assemble(F).dat.data
```

Results in a compilation error.  This is on either fix-thetis or master in coffee, master everything else.

### PR DESCRIPTION
**Pull request by FabioLuporini**
Most changes are actually just more sensible variable names and fixes to the doc. 
In essence, this PR: 
- adds a tree visitor for performing reachability analysis
- uses reachability analysis for correct code motion during CSE

fixes #74

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -15,8 +15,8 @@ transformations:
   alignment, a transformation for enhancing the effectiveness of compiler
   auto-vectorisation.
 
-COFFEE currently has one user, Firedrake, an automated system for the solution of
+COFFEE currently has one user, Firedrake, an automated system for the resolution of
```
Comment (miklos1) [inline README.rst]: Really?

File: coffee/utils.py [hunk 2]
```diff
@@ -358,6 +358,19 @@ def ldanalysis(node, key='default', value='default'):
     return lda
 
 
+def ranalysis(node, decls=None):
```
Comment (miklos1) [inline coffee/utils.py]: `r` doesn't tell much, maybe `reachability_analysis`?

File: coffee/visitors/inspectors.py [hunk 3]
```diff
@@ -554,6 +495,56 @@ def visit_Decl(self, o, ret=None, scope=LOCAL, *args, **kwargs):
         return ret
 
 
+class SymbolVisibility(Visitor):
+
+    @classmethod
+    def default_retval(cls):
+        return defaultdict(list), []
+
+    """
```
Comment (miklos1) [inline coffee/visitors/inspectors.py]: Is this still recognised as a docstring, despite not being the first thing in the class?

Comment (FabioLuporini) [inline coffee/visitors/inspectors.py]: It worked with the other visitors I seem to recall

Comment (miklos1) [inline coffee/visitors/inspectors.py]: OK

**Pull request merged**